ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron/Ledger.hs

```diff
@@ -0,0 +1,299 @@
+  -- it becomes active (i.e., that delegation should be applied /in/ that slot)
+  -- i.e., that delegate is allowed to issue a block in that very same slot.
+  anachronisticProtocolLedgerView cfg (ByronLedgerState ls ss) slot =
+      case History.find slot ss of
+        Just sb -> Right (toPBftLedgerView <$> sb) -- Case (A)
+        Nothing -- Case (B), (C) or (D)
+          | slot <  At maxLo -> Left TooFarBehind -- lower bound is inclusive
+          | slot >= At maxHi -> Left TooFarAhead  -- upper bound is exclusive
+          | otherwise        -> Right $ toPBftLedgerView <$>
+
+              let toApply :: Seq D.Sched.ScheduledDelegation
+                  future  :: Seq D.Sched.ScheduledDelegation
+                  (toApply, future) = splitScheduledDelegations slot $
+                                        getScheduledDelegations ls
+
+                  lo, hi :: SlotNo
+                  lo = case (toApply, History.toSequence ss) of
+                         (_ Seq.Lazy.:|> upd, _) ->
+                           fromByronSlot (D.Sched.sdSlot upd)
+                         (Seq.Lazy.Empty, _ Seq.:|> prev) ->
+                           SB.sbUpper prev
+                         (Seq.Lazy.Empty, Seq.Empty) ->
+                           genesisSlotNo
+                  hi = case future of
+                         Seq.Lazy.Empty ->
+                           maxHi
+                         upd Seq.Lazy.:<| _ ->
+                           fromByronSlot (D.Sched.sdSlot upd)
+
+              in SB.bounded lo hi $ applyScheduledDelegations toApply dsNow
+    where
+      SecurityParam k = pbftSecurityParam . pbftParams $ cfg
+
+      dsNow :: Delegation.Map
+      dsNow = getDelegationMap ls
+
+      now, maxHi, maxLo :: SlotNo
+      now   = fromByronSlot $ CC.cvsLastSlot ls
+      maxLo = SlotNo $ if 2 * k > unSlotNo now
+                        then 0
+                        else unSlotNo now - (2 * k)
+      maxHi = SlotNo $ unSlotNo now + (2 * k)
+
+{-------------------------------------------------------------------------------
+  Auxiliary
+-------------------------------------------------------------------------------}
+
+-- | Mark computation as validation error free
+--
+-- Given a 'BlockValidationMode' of 'NoBlockValidation', a call to
+-- 'applyByronBlock' shouldn't fail since the ledger layer won't be performing
+-- any block validation checks. However, because 'applyByronBlock' can fail in
+-- the event it is given a 'BlockValidationMode' of 'BlockValidation', it still
+-- /looks/ like it can fail (since its type doesn't change based on the
+-- 'ValidationMode') and we must still treat it as such.
+validationErrorImpossible :: forall err a. Except err a -> a
+validationErrorImpossible = cantBeError . runExcept
+  where
+    cantBeError :: Either err a -> a
+    cantBeError (Left  _) = error "validationErrorImpossible: unexpected error"
+    cantBeError (Right a) = a
+
+-- | Split scheduled delegations into past and future
```

Comment:
Big ass comment: we deviate from the spec here! In `applyExtLedgerState`, we _first_ apply updates from the body, and _then_ do the PBFT signer check. What this means is that for us delegation updates happen _before_ the signer is checked, and so an update scheduled for slot `n` actually applies in slot `n`. In the spec however the signer is checked _before_ the body and so an update scheduled for slot `n` _really_ applies at slot `n+1`.

Reply:
This comment is now part of `applyExtLedgerState`.